codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java

```diff
@@ -56,22 +57,25 @@ public DefaultHttp2LocalFlowController(Http2Connection connection,
         this.frameWriter = checkNotNull(frameWriter, "frameWriter");
         windowUpdateRatio(windowUpdateRatio);
 
-        // Add a flow state for the connection.
-        final Http2Stream connectionStream = connection.connectionStream();
-        connectionStream.setProperty(FlowState.class, new FlowState(connectionStream, initialWindowSize));
-
         // Register for notification of new streams.
         connection.addListener(new Http2ConnectionAdapter() {
             @Override
-            public void onStreamAdded(Http2Stream stream) {
-                stream.setProperty(FlowState.class, new FlowState(stream, 0));
-            }
-
-            @Override
             public void onStreamActive(Http2Stream stream) {
+                DefaultFlowState state = state(stream);
+                state.streamActivated();
                 // Need to be sure the stream's initial window is adjusted for SETTINGS
                 // frames which may have been exchanged while it was in IDLE
-                state(stream).window(initialWindowSize);
+                state.window(initialWindowSize);
```

Comment:
Potentially.  I just want to make sure we want to support this concept first.  See https://github.com/netty/netty/pull/3616#discussion_r28251536 for more explanation.  I'll try to prototype it.